src/components/vm/consoles/serial.jsx

```diff
@@ -0,0 +1,179 @@
+                oldSpawnArgs.some((arg, index) => arg !== newSpawnArgs[index]))
+                return true;
+            if (this.props.state.connected && !this.state.channel)
+                return true;
+            if (!this.props.state.connected && this.state.channel)
+                return true;
+            return false;
+        };
+
+        if (channel_needs_update())
+            this.updateChannel(this.props.spawnArgs);
+    }
+
+    updateChannel() {
+        if (this.state.channel)
+            this.state.channel.close();
+
+        if (this.props.state.connected) {
+            const opts = {
+                payload: "stream",
+                spawn: this.props.spawnArgs,
+                pty: true,
+            };
+            if (this.props.connectionName == "system")
+                opts.superuser = "try";
+            const channel = cockpit.channel(opts);
+            this.setState({ channel });
+        } else {
+            this.setState({ channel: null });
+        }
+    }
+
+    render () {
+        const { state } = this.props;
+
+        const pid = this.props.vmName + "-terminal";
+        let t;
+        if (!state.connected) {
+            t = (
+                <EmptyState>
+                    <EmptyStateBody>{_("Disconnected")}</EmptyStateBody>
+                    <EmptyStateFooter>
+                        <Button variant="primary" onClick={() => state.setConnected(true)}>
+                            {_("Connect")}
+                        </Button>
+                    </EmptyStateFooter>
+                </EmptyState>
+            );
+        } else if (!this.state.channel) {
+            t = <span>{_("Loading...")}</span>;
+        } else {
+            t = (
+                <Terminal
+                    refName={this.props.vmName}
+                    channel={this.state.channel}
+                    parentId={pid}
+                />
+            );
+        }
+        return (
+            <div id={pid} className="vm-terminal vm-console-serial">
+                {t}
+            </div>
+        );
+    }
+}
+
+SerialActive.propTypes = {
+    connectionName: PropTypes.string.isRequired,
+    vmName: PropTypes.string.isRequired,
+    spawnArgs: PropTypes.array.isRequired,
+};
+
+export const SerialInactive = ({ vm }) => {
+    return (
+        <EmptyState>
+            <EmptyStateBody>
+                {_("Start the virtual machine to access the console")}
+            </EmptyStateBody>
+        </EmptyState>
+    );
+};
+
+export const SerialMissing = ({ vm, onAddErrorNotification }) => {
+    const [inProgress, setInProgress] = useState(false);
+
+    function add_serial() {
+        setInProgress(true);
+        domainAttachSerialConsole(vm)
+                .catch(ex => onAddErrorNotification({
+                    text: cockpit.format(_("Failed to add text console to VM $0"), vm.name),
+                    detail: ex.message,
+                    resourceId: vm.id,
```

Comment:
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2008-6fca6491-20250714-153804-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)